public abstract class kotlinx.coroutines.DispatchedTask extends kotlinx.coroutines.scheduling.Task
{
    public int resumeMode;

    public void <init>(int)
    {
        int $i0;
        kotlinx.coroutines.DispatchedTask r0;

        r0 := @this: kotlinx.coroutines.DispatchedTask;

        $i0 := @parameter0: int;

        specialinvoke r0.<kotlinx.coroutines.scheduling.Task: void <init>()>();

        r0.<kotlinx.coroutines.DispatchedTask: int resumeMode> = $i0;

        return;
    }

    public void cancelCompletedResult$kotlinx_coroutines_core(java.lang.Object, java.lang.Throwable)
    {
        java.lang.Throwable $r2;
        kotlinx.coroutines.DispatchedTask r0;
        java.lang.Object $r1;

        r0 := @this: kotlinx.coroutines.DispatchedTask;

        $r1 := @parameter0: java.lang.Object;

        $r2 := @parameter1: java.lang.Throwable;

        return;
    }

    public abstract kotlin.coroutines.Continuation getDelegate$kotlinx_coroutines_core();

    public java.lang.Throwable getExceptionalResult$kotlinx_coroutines_core(java.lang.Object)
    {
        java.lang.Throwable r3;
        kotlinx.coroutines.DispatchedTask r0;
        kotlinx.coroutines.CompletedExceptionally $r2;
        java.lang.Object $r1;
        boolean $z0;

        r0 := @this: kotlinx.coroutines.DispatchedTask;

        $r1 := @parameter0: java.lang.Object;

        $z0 = $r1 instanceof kotlinx.coroutines.CompletedExceptionally;

        if $z0 == 0 goto label1;

        $r2 = (kotlinx.coroutines.CompletedExceptionally) $r1;

        goto label2;

     label1:
        $r2 = null;

     label2:
        if $r2 == null goto label3;

        r3 = $r2.<kotlinx.coroutines.CompletedExceptionally: java.lang.Throwable cause>;

        return r3;

     label3:
        return null;
    }

    public java.lang.Object getSuccessfulResult$kotlinx_coroutines_core(java.lang.Object)
    {
        kotlinx.coroutines.DispatchedTask r0;
        java.lang.Object $r1;

        r0 := @this: kotlinx.coroutines.DispatchedTask;

        $r1 := @parameter0: java.lang.Object;

        return $r1;
    }

    public final void handleFatalException(java.lang.Throwable, java.lang.Throwable)
    {
        java.lang.Throwable $r1, $r2;
        kotlinx.coroutines.DispatchedTask r0;
        kotlin.coroutines.Continuation $r6;
        kotlin.coroutines.CoroutineContext $r7;
        java.lang.StringBuilder $r4;
        java.lang.String $r5;
        kotlinx.coroutines.CoroutinesInternalError $r3;

        r0 := @this: kotlinx.coroutines.DispatchedTask;

        $r1 := @parameter0: java.lang.Throwable;

        $r2 := @parameter1: java.lang.Throwable;

        if $r1 != null goto label1;

        if $r2 != null goto label1;

        return;

     label1:
        if $r1 == null goto label2;

        if $r2 == null goto label2;

        staticinvoke <kotlin.ExceptionsKt: void addSuppressed(java.lang.Throwable,java.lang.Throwable)>($r1, $r2);

     label2:
        if $r1 != null goto label3;

        $r1 = $r2;

        goto label3;

     label3:
        $r3 = new kotlinx.coroutines.CoroutinesInternalError;

        $r4 = new java.lang.StringBuilder;

        specialinvoke $r4.<java.lang.StringBuilder: void <init>()>();

        $r4 = virtualinvoke $r4.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.String)>("Fatal exception in coroutines machinery for ");

        $r4 = virtualinvoke $r4.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.Object)>(r0);

        $r4 = virtualinvoke $r4.<java.lang.StringBuilder: java.lang.StringBuilder append(java.lang.String)>(". Please read KDoc to \'handleFatalException\' method and report this incident to maintainers");

        $r5 = virtualinvoke $r4.<java.lang.StringBuilder: java.lang.String toString()>();

        staticinvoke <kotlin.jvm.internal.Intrinsics: void checkNotNull(java.lang.Object)>($r1);

        specialinvoke $r3.<kotlinx.coroutines.CoroutinesInternalError: void <init>(java.lang.String,java.lang.Throwable)>($r5, $r1);

        $r6 = virtualinvoke r0.<kotlinx.coroutines.DispatchedTask: kotlin.coroutines.Continuation getDelegate$kotlinx_coroutines_core()>();

        $r7 = interfaceinvoke $r6.<kotlin.coroutines.Continuation: kotlin.coroutines.CoroutineContext getContext()>();

        $r1 = (java.lang.Throwable) $r3;

        staticinvoke <kotlinx.coroutines.CoroutineExceptionHandlerKt: void handleCoroutineException(kotlin.coroutines.CoroutineContext,java.lang.Throwable)>($r7, $r1);

        return;
    }

    public final void run()
    {
        java.lang.Throwable r3, $r12, $r16, $r21, $r22;
        kotlinx.coroutines.DispatchedTask r0;
        kotlin.coroutines.Continuation $r4;
        kotlinx.coroutines.Job$Key r13;
        kotlin.coroutines.jvm.internal.CoroutineStackFrame r19;
        kotlinx.coroutines.Job r9;
        kotlin.Result$Companion $r18;
        kotlin.coroutines.CoroutineContext$Element $r15;
        kotlinx.coroutines.internal.DispatchedContinuation $r5;
        java.lang.AssertionError $r23;
        int i0;
        boolean $z0;
        kotlinx.coroutines.internal.Symbol $r8;
        java.util.concurrent.CancellationException $r17;
        kotlinx.coroutines.scheduling.TaskContext r2;
        kotlin.coroutines.CoroutineContext $r7, $r10;
        kotlin.Unit $r20;
        kotlin.coroutines.CoroutineContext$Key r14;
        java.lang.Object $r6, $r11;
        kotlinx.coroutines.UndispatchedCoroutine r1;

        r0 := @this: kotlinx.coroutines.DispatchedTask;

        $z0 = staticinvoke <kotlinx.coroutines.DebugKt: boolean getASSERTIONS_ENABLED()>();

        if $z0 == 0 goto label04;

        i0 = r0.<kotlinx.coroutines.DispatchedTask: int resumeMode>;

        if i0 == -1 goto label01;

        $z0 = 1;

        goto label02;

     label01:
        $z0 = 0;

     label02:
        if $z0 == 0 goto label03;

        goto label04;

     label03:
        $r23 = new java.lang.AssertionError;

        specialinvoke $r23.<java.lang.AssertionError: void <init>()>();

        throw $r23;

     label04:
        r2 = r0.<kotlinx.coroutines.DispatchedTask: kotlinx.coroutines.scheduling.TaskContext taskContext>;

        r3 = null;

     label05:
        $r4 = virtualinvoke r0.<kotlinx.coroutines.DispatchedTask: kotlin.coroutines.Continuation getDelegate$kotlinx_coroutines_core()>();

        staticinvoke <kotlin.jvm.internal.Intrinsics: void checkNotNull(java.lang.Object,java.lang.String)>($r4, "null cannot be cast to non-null type kotlinx.coroutines.internal.DispatchedContinuation<T of kotlinx.coroutines.DispatchedTask>");

        $r5 = (kotlinx.coroutines.internal.DispatchedContinuation) $r4;

        $r4 = $r5.<kotlinx.coroutines.internal.DispatchedContinuation: kotlin.coroutines.Continuation continuation>;

        $r6 = $r5.<kotlinx.coroutines.internal.DispatchedContinuation: java.lang.Object countOrElement>;

        $r7 = interfaceinvoke $r4.<kotlin.coroutines.Continuation: kotlin.coroutines.CoroutineContext getContext()>();

        $r6 = staticinvoke <kotlinx.coroutines.internal.ThreadContextKt: java.lang.Object updateThreadContext(kotlin.coroutines.CoroutineContext,java.lang.Object)>($r7, $r6);

        $r8 = <kotlinx.coroutines.internal.ThreadContextKt: kotlinx.coroutines.internal.Symbol NO_THREAD_ELEMENTS>;

     label06:
        r9 = null;

        if $r6 == $r8 goto label09;

     label07:
        r1 = staticinvoke <kotlinx.coroutines.CoroutineContextKt: kotlinx.coroutines.UndispatchedCoroutine updateUndispatchedCompletion(kotlin.coroutines.Continuation,kotlin.coroutines.CoroutineContext,java.lang.Object)>($r4, $r7, $r6);

     label08:
        goto label10;

     label09:
        r1 = null;

     label10:
        $r10 = interfaceinvoke $r4.<kotlin.coroutines.Continuation: kotlin.coroutines.CoroutineContext getContext()>();

        $r11 = virtualinvoke r0.<kotlinx.coroutines.DispatchedTask: java.lang.Object takeState$kotlinx_coroutines_core()>();

        $r12 = virtualinvoke r0.<kotlinx.coroutines.DispatchedTask: java.lang.Throwable getExceptionalResult$kotlinx_coroutines_core(java.lang.Object)>($r11);

     label11:
        if $r12 != null goto label17;

     label12:
        i0 = r0.<kotlinx.coroutines.DispatchedTask: int resumeMode>;

        $z0 = staticinvoke <kotlinx.coroutines.DispatchedTaskKt: boolean isCancellableMode(int)>(i0);

     label13:
        if $z0 == 0 goto label17;

     label14:
        r13 = <kotlinx.coroutines.Job: kotlinx.coroutines.Job$Key Key>;

        r14 = (kotlin.coroutines.CoroutineContext$Key) r13;

        $r15 = interfaceinvoke $r10.<kotlin.coroutines.CoroutineContext: kotlin.coroutines.CoroutineContext$Element get(kotlin.coroutines.CoroutineContext$Key)>(r14);

        r9 = (kotlinx.coroutines.Job) $r15;

     label15:
        goto label17;

     label16:
        $r16 := @caughtexception;

        goto label45;

     label17:
        if r9 == null goto label31;

     label18:
        $z0 = interfaceinvoke r9.<kotlinx.coroutines.Job: boolean isActive()>();

     label19:
        if $z0 != 0 goto label31;

     label20:
        $r17 = interfaceinvoke r9.<kotlinx.coroutines.Job: java.util.concurrent.CancellationException getCancellationException()>();

        $r12 = (java.lang.Throwable) $r17;

        virtualinvoke r0.<kotlinx.coroutines.DispatchedTask: void cancelCompletedResult$kotlinx_coroutines_core(java.lang.Object,java.lang.Throwable)>($r11, $r12);

        $r18 = <kotlin.Result: kotlin.Result$Companion Companion>;

        $z0 = staticinvoke <kotlinx.coroutines.DebugKt: boolean getRECOVER_STACK_TRACES()>();

     label21:
        if $z0 == 0 goto label28;

     label22:
        $z0 = $r4 instanceof kotlin.coroutines.jvm.internal.CoroutineStackFrame;

     label23:
        if $z0 != 0 goto label24;

        goto label28;

     label24:
        $r12 = (java.lang.Throwable) $r17;

     label25:
        r19 = (kotlin.coroutines.jvm.internal.CoroutineStackFrame) $r4;

        $r12 = staticinvoke <kotlinx.coroutines.internal.StackTraceRecoveryKt: java.lang.Throwable access$recoverFromStackFrame(java.lang.Throwable,kotlin.coroutines.jvm.internal.CoroutineStackFrame)>($r12, r19);

     label26:
        goto label29;

     label27:
        $r16 := @caughtexception;

        goto label45;

     label28:
        $r12 = (java.lang.Throwable) $r17;

     label29:
        $r11 = staticinvoke <kotlin.ResultKt: java.lang.Object createFailure(java.lang.Throwable)>($r12);

        $r11 = staticinvoke <kotlin.Result: java.lang.Object constructor-impl(java.lang.Object)>($r11);

        interfaceinvoke $r4.<kotlin.coroutines.Continuation: void resumeWith(java.lang.Object)>($r11);

     label30:
        goto label35;

     label31:
        if $r12 == null goto label34;

     label32:
        $r18 = <kotlin.Result: kotlin.Result$Companion Companion>;

        $r11 = staticinvoke <kotlin.ResultKt: java.lang.Object createFailure(java.lang.Throwable)>($r12);

        $r11 = staticinvoke <kotlin.Result: java.lang.Object constructor-impl(java.lang.Object)>($r11);

        interfaceinvoke $r4.<kotlin.coroutines.Continuation: void resumeWith(java.lang.Object)>($r11);

     label33:
        goto label35;

     label34:
        $r18 = <kotlin.Result: kotlin.Result$Companion Companion>;

        $r11 = virtualinvoke r0.<kotlinx.coroutines.DispatchedTask: java.lang.Object getSuccessfulResult$kotlinx_coroutines_core(java.lang.Object)>($r11);

        $r11 = staticinvoke <kotlin.Result: java.lang.Object constructor-impl(java.lang.Object)>($r11);

        interfaceinvoke $r4.<kotlin.coroutines.Continuation: void resumeWith(java.lang.Object)>($r11);

     label35:
        $r20 = <kotlin.Unit: kotlin.Unit INSTANCE>;

     label36:
        if r1 == null goto label39;

     label37:
        $z0 = virtualinvoke r1.<kotlinx.coroutines.UndispatchedCoroutine: boolean clearThreadContext()>();

     label38:
        if $z0 == 0 goto label40;

     label39:
        staticinvoke <kotlinx.coroutines.internal.ThreadContextKt: void restoreThreadContext(kotlin.coroutines.CoroutineContext,java.lang.Object)>($r7, $r6);

     label40:
        $r18 = <kotlin.Result: kotlin.Result$Companion Companion>;

        interfaceinvoke r2.<kotlinx.coroutines.scheduling.TaskContext: void afterTask()>();

        $r20 = <kotlin.Unit: kotlin.Unit INSTANCE>;

        $r6 = staticinvoke <kotlin.Result: java.lang.Object constructor-impl(java.lang.Object)>($r20);

     label41:
        goto label55;

     label42:
        $r21 := @caughtexception;

        goto label54;

     label43:
        $r16 := @caughtexception;

        goto label45;

     label44:
        $r16 := @caughtexception;

     label45:
        if r1 == null goto label48;

     label46:
        $z0 = virtualinvoke r1.<kotlinx.coroutines.UndispatchedCoroutine: boolean clearThreadContext()>();

     label47:
        if $z0 == 0 goto label49;

     label48:
        staticinvoke <kotlinx.coroutines.internal.ThreadContextKt: void restoreThreadContext(kotlin.coroutines.CoroutineContext,java.lang.Object)>($r7, $r6);

     label49:
        throw $r16;

     label50:
        $r22 := @caughtexception;

        r3 = $r22;

     label51:
        $r18 = <kotlin.Result: kotlin.Result$Companion Companion>;

        interfaceinvoke r2.<kotlinx.coroutines.scheduling.TaskContext: void afterTask()>();

        $r20 = <kotlin.Unit: kotlin.Unit INSTANCE>;

        $r6 = staticinvoke <kotlin.Result: java.lang.Object constructor-impl(java.lang.Object)>($r20);

     label52:
        goto label55;

     label53:
        $r21 := @caughtexception;

     label54:
        $r18 = <kotlin.Result: kotlin.Result$Companion Companion>;

        $r6 = staticinvoke <kotlin.ResultKt: java.lang.Object createFailure(java.lang.Throwable)>($r21);

        $r6 = staticinvoke <kotlin.Result: java.lang.Object constructor-impl(java.lang.Object)>($r6);

     label55:
        $r12 = staticinvoke <kotlin.Result: java.lang.Throwable exceptionOrNull-impl(java.lang.Object)>($r6);

        virtualinvoke r0.<kotlinx.coroutines.DispatchedTask: void handleFatalException(java.lang.Throwable,java.lang.Throwable)>(r3, $r12);

        return;

        catch java.lang.Throwable from label05 to label06 with label50;
        catch java.lang.Throwable from label07 to label08 with label50;
        catch java.lang.Throwable from label10 to label11 with label44;
        catch java.lang.Throwable from label12 to label13 with label16;
        catch java.lang.Throwable from label14 to label15 with label16;
        catch java.lang.Throwable from label18 to label19 with label44;
        catch java.lang.Throwable from label20 to label21 with label44;
        catch java.lang.Throwable from label22 to label23 with label27;
        catch java.lang.Throwable from label24 to label25 with label27;
        catch java.lang.Throwable from label25 to label26 with label43;
        catch java.lang.Throwable from label28 to label30 with label43;
        catch java.lang.Throwable from label32 to label33 with label43;
        catch java.lang.Throwable from label34 to label36 with label43;
        catch java.lang.Throwable from label37 to label38 with label50;
        catch java.lang.Throwable from label39 to label40 with label50;
        catch java.lang.Throwable from label40 to label41 with label42;
        catch java.lang.Throwable from label46 to label47 with label50;
        catch java.lang.Throwable from label48 to label50 with label50;
        catch java.lang.Throwable from label51 to label52 with label53;
    }

    public abstract java.lang.Object takeState$kotlinx_coroutines_core();
}
